VSRAD.Syntax/Core/DocumentFactory.cs

```diff
@@ -0,0 +1,141 @@
+        }
+
+        public IDocument GetOrCreateDocument(string path)
+        {
+            if (_documents.TryGetValue(path, out var document))
+                return document;
+
+            var contentType = _contentTypeManager.DetermineContentType(path);
+            if (contentType == null)
+                return null;
+
+            var textDocument = _serviceProvider
+                .TextDocumentFactoryService
+                .CreateAndLoadTextDocument(path, contentType);
+
+            return CreateDocument(textDocument, (lexer, parser) => new InvisibleDocument(textDocument, lexer, parser));
+        }
+
+        public IDocument GetOrCreateDocument(ITextBuffer buffer)
+        {
+            if (buffer.Properties.TryGetProperty(typeof(IDocument), out IDocument document))
+                return document;
+
+            var textDocument = buffer.Properties.GetProperty<ITextDocument>(typeof(ITextDocument));
+            if (_documents.TryGetValue(textDocument.FilePath, out document)
+                && document is InvisibleDocument invisibleDocument)
+            {
+                document = invisibleDocument.ToVisibleDocument(textDocument);
+                ObserveDocument(document, textDocument);
+            }
+            else
+            {
+                document = CreateDocument(textDocument, (lexer, parser) => new Document(textDocument, lexer, parser));
+            }
+
+            DocumentCreated?.Invoke(document);
+            return document;
+        }
+
+        private IDocument CreateDocument(ITextDocument textDocument, Func<ILexer, IParser, IDocument> creator)
+        {
+            var lexerParser = GetLexerParser(textDocument.TextBuffer.ContentType);
+            if (!lexerParser.HasValue) return null;
+
+            var lexer = lexerParser.Value.lexer;
+            var parser = lexerParser.Value.parser;
+
+            var document = creator(lexer, parser);
+            ObserveDocument(document, textDocument);
+
+            return document;
+        }
+
+        private void ObserveDocument(IDocument document, ITextDocument textDocument)
+        {
+            document.DocumentRenamed += DocumentRenamed;
+            textDocument.TextBuffer.Properties.AddProperty(typeof(IDocument), document);
+
+            _documents.Add(document.Path, document);
+        }
+
+        private (ILexer lexer, IParser parser)? GetLexerParser(IContentType contentType)
+        {
+            if (contentType == _contentTypeManager.Asm1ContentType)
+                return (new AsmLexer(), new Asm1Parser(this, _instructionManager));
+            else if (contentType == _contentTypeManager.Asm2ContentType)
+                return (new Asm2Lexer(), new Asm2Parser(this, _instructionManager));
+            else if (contentType == _contentTypeManager.AsmDocContentType)
+                return (new AsmDocLexer(), new AsmDocParser());
+
+            else return null;
```

Comment:
Better to rewrite using switch statement?

Reply:
You're right, fixed in 487994a